src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/caseload/CaCaseloadService.kt

```diff
@@ -0,0 +1,492 @@
+
+    val missingDates = listOf(
+      IN_PROGRESS,
+      SUBMITTED,
+      APPROVED,
+      NOT_STARTED,
+    ).contains(status) &&
+      noReleaseDates
+    val startDateInPast = licenceStartDate != null && status == APPROVED && licenceStartDate.isBefore(today)
+
+    return missingDates || startDateInPast
+  }
+
+  private fun selectReleaseDate(nomisRecord: Prisoner): LocalDate? {
+    return nomisRecord.confirmedReleaseDate ?: nomisRecord.conditionalReleaseDate
+  }
+
+  private fun getPrisonersApproachingRelease(
+    prisonCaseload: List<String>?,
+    overrideClock: Clock? = null,
+  ): Page<CaseloadItem>? {
+    val now = overrideClock ?: clock
+    val weeksToAdd: Long = 4
+    val today = LocalDate.now(now)
+    val todayPlusFourWeeks = LocalDate.now(now).plusWeeks(weeksToAdd)
+    return prisonCaseload?.toSet()?.let {
+      caseloadService.getPrisonersByReleaseDate(
+        today,
+        todayPlusFourWeeks,
+        it,
+        page = 0,
+      )
+    }
+  }
+
+  fun findLatestLicenceSummary(licences: List<LicenceSummary>): LicenceSummary? {
+    if (licences.size == 1) {
+      return licences[0]
+    }
+    if (licences.any { it.licenceStatus == TIMED_OUT }) {
+      return licences.find { it.licenceStatus != TIMED_OUT }
+    }
+
+    return licences.find { (it.licenceStatus == SUBMITTED) || (it.licenceStatus == IN_PROGRESS) }
+  }
+
+  private fun applySearch(searchString: String?, cases: List<CaCase>): List<CaCase> {
+    if (searchString == null) {
+      return cases
+    }
+    val term = searchString.lowercase()
+    return cases.filter {
+      it.name.lowercase().contains(term) ||
+        it.prisonerNumber.lowercase().contains(term) ||
+        it.probationPractitioner?.name?.lowercase()?.contains(term) ?: false
+    }
+  }
+
+  private fun applySort(cases: List<CaCase>, view: String): List<CaCase> =
+    if (view == "prison") cases.sortedBy { it.releaseDate } else cases.sortedByDescending { it.releaseDate }
+
+  private fun pairNomisRecordsWithDelius(prisoners: List<CaseloadItem>): List<ManagedCase> {
+    val caseloadNomisIds = prisoners
+      .map { it.prisoner.prisonerNumber!! }
+
+    val deliusRecords = probationSearchApiClient.searchForPeopleByNomsNumber(caseloadNomisIds)
+
+    return prisoners
+      .map { (prisoner, cvl) ->
+        val deliusRecord = deliusRecords.find { d -> d.otherIds.nomsNumber == prisoner.prisonerNumber }
+        if (deliusRecord != null) {
+          return@map ManagedCase(
+            nomisRecord = prisoner,
+            cvlFields = cvl,
+            deliusRecord = DeliusRecord(
+              deliusRecord,
+              ManagedOffenderCrn(
+                staff = deliusRecord.offenderManagers.find { om -> om.active }?.staffDetail,
+              ),
+            ),
+          )
+        }
+        return@map ManagedCase(nomisRecord = prisoner, cvlFields = cvl)
+      }
+  }
+
+  private fun mapCasesToComs(casesToMap: List<CaCase>): List<CaCase> {
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏This is quite confusing as to what is happening at first glance - here we are creating the caseload from the groupings with the relevant COMs where they exist or not - only by going through it bit by bit, you understand what is happening. Brain starting to hurt!

Reply:
This bit is on me (the API is a recreating of the refactor that's happened in the UI)
In short, this is to reduce API calls and/or the size of said API calls to Delius. 
For cases where we have the COM staff code, we will also have their name, so don't need to call to Delius
For cases where we have the COM username, we should use that username to get the staff code and name from Delius
For cases where we have neither, we need to use the NOMIS ID to get the offender record, and get the COM details from that

Reply:
To make it clear I have added comments on it